Cleanup MEE-CBC

amaloz opened the pull request:
This PR cleans up the `MEE_CBC.cry` implementation as follows:
* It generalizes the scheme to work for any CBC schemes and HMACs, moving the particular instantiation using AES128-CBC and HMAC-SHA1 to `Instantiations/MEE_CBC_SHA1.cry`.
* It improves the documentation and otherwise aims to make the MEE-CBC implementation more up-to-snuff with the gold standard requirements.

amaloz committed:
MEE-CBC: First step towards cleaning up implementation

This first attempt hardcodes a specific CBC and HMAC, and using those cleans up
the implementation. Still need documentation and potentially abstracting out
the particular CBC and HMAC implementation.

amaloz committed:
MEE-CBC: Add documentation

amaloz committed:
MEE-CBC: Generalize the MEE-CBC construction

Before, the MEE-CBC construction was fixed to use HMAC-SHA1 and AES128-CBC.
This commit turns MEE-CBC from an instantiation to a specification, allowing
various instantiations of MEE-CBC for different HMAC and CBC instantiations.
In order to do this we introduce interfaces for HMAC and CBC.

amaloz committed:
MEE-CBC: Fix padding

The padding algorithm in the spec says to pad with `p` bytes, but how the padding
is actually used for MEE requires `p+1` bytes. This commit fixes that issue.

amaloz committed:
MEE-CBC: Add missing whitespace

amaloz committed:
MEE-CBC: Add missing documentation

marsella commented on Primitive/Symmetric/Cipher/Authenticated/MEE_CBC/Specification.cry:
```diff
@@ -0,0 +1,159 @@
+ * References:
+ * [TLS1.2] T. Dierks and E. Rescorla. The Transport Layer Security (TLS)
+   Protocol Version 1.2. RFC 5246, August 2008.
+   @see http://www.ietf.org/rfc/rfc5246.txt.
```
minor syntax request: add `*`s to line up with the ones on the other lines
```suggestion
 *     Protocol Version 1.2. RFC 5246, August 2008.
 *     @see http://www.ietf.org/rfc/rfc5246.txt.
```

marsella commented on Primitive/Symmetric/Cipher/Block/Modes/CBCInterface.cry:
thought: This interface would probably be identical for many modes of operation, but it's nice to have it be documented as mode-specific to make it harder to accidentally use an unsuitable mode.

marsella reviewed (commented):
This mostly looks good. Main blocker to approving is clarification on how to read the spec's description of mee-cbc -- are there any other sections that give context on how to read that bit?

amaloz commented on Primitive/Symmetric/Cipher/Authenticated/MEE_CBC/Specification.cry:
```diff
@@ -0,0 +1,159 @@
+    m = take`{n} ct
+    t = take`{HMAC::TagSize, p + 1} (drop`{n} ct)
+    padding = drop`{n + HMAC::TagSize} ct
+    good = padding == ((repeat `p) : [p + 1][8])
```
Oh I like that!

amaloz commented on Primitive/Symmetric/Cipher/Authenticated/MEE_CBC/Specification.cry:
```diff
@@ -0,0 +1,159 @@
+    good = padding == ((repeat `p) : [p + 1][8])
+
+/**
+ * Validate that padding inverts as expected.
```
Is there a way to write a test against a particular block size? That is, the padding is a function in part on the block size, but that isn't specified until the CBC interface is instantiated. So depending on how its instantiated the test could fail or not (unless there's a way to pin the test to work only for a particular block size, but for the example in the spec we don't have an instantiation that matches it, since it uses a block size of 64 bits).

marsella replied:
Oh, hm. I see what you mean. I tried a bunch of stuff but I keep running into the issue that there's no relationship between `TypeSize` and `BlockSize` so it's hard to come up with a formula for a valid `p`.

Here is the best workaround I came up with:
1. Define the generic property, like you have;
2. Define a special test-case function that calls the property but adds numeric constraint guards to define the cases we actually have instantiations for;
3. Call the special function in the docstring, instead of the generic property
```
/**
 * Validate that padding inverts as expected.
 * ```repl
 * :prove invertTestCases
 * ```
 */
paddingInverts : {n, p} (fin n, 8 >= width p, 8 * (1 + (HMAC::TagSize + n + p)) % CBC::BlockSize == 0)
  => [n][8] -> [HMAC::TagSize][8] -> Bit
property paddingInverts msg tag = optTest (\x -> x == (msg, tag)) should_invert where
    should_invert = unpad`{n, p} (pad`{n, p} msg tag)

invertTestCases : [61][8] -> [HMAC::TagSize][8] -> Bit
invertTestCases msg tag
  // With AES128 and SHA1
  | (CBC::BlockSize == 128, HMAC::TagSize == 20) => paddingInverts`{61, 14} msg tag
  | () => False
```

I put `False` in the otherwise case because that'll force implementers to add tests if they add an instantiation that has a different block / tag size.

What do you think of that? It's very specialized and picking so many parameters makes it a pretty narrow test.

marsella replied:
Oh sorry I just saw your approach. I think that is also great; we can add all this funky constraint guard stuff if we ever add more instantiations.

amaloz committed:
MEE-CBC: Address reviewer comments

This commit cleans up a bunch of the documentation, alongside other minor
niceties raised during the review process.

amaloz committed:
MEE-CBC: Update property checks

This adds a check docstring to validate the properties on a specific
parameter instantiation.

amaloz commented:
I think this is ready for a re-review.

marsella reviewed (approved):
Thank you, these changes look good to me!
